### PR TITLE
Get servo working on win32

### PR DESCRIPTION
**Pull request by vvuk**
This depends on a bunch of other PRs; full instructions and outstanding PRs are at https://hackpad.com/Servo-on-Windows-C1LPcI2bP25

Most of the servo-side work is just some glue bits for fonts and some build changes.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/7878)

### CODE REVIEW

**Comment by nox** [comment]
@vvuk This needs a rebase.

File: components/util/workqueue.rs [hunk 1]
```diff
@@ -151,9 +169,8 @@ impl<QueueData: Sync, WorkData: Send> WorkerThread<QueueData, WorkData> {
                                     }
                                 }
 
-                                unsafe {
-                                    usleep(back_off_sleep as u32);
-                                }
+                                sleep_microseconds(back_off_sleep);
```
Comment (eefriedman) [inline components/util/workqueue.rs]: I think you can use `std::thread::sleep` here.

File: python/servo/command_base.py [hunk 2]
```diff
@@ -313,6 +327,9 @@ def build_env(self, gonk=False, hosts_file_path=None):
             if subprocess.call(['which', 'ld.gold'], stdout=PIPE, stderr=PIPE) == 0:
                 env['RUSTC'] = path.join(self.context.topdir, 'etc', 'rustc-with-gold')
 
+        # This sometimes ends up as unicode
+        env["PATH"] = env["PATH"].encode('ascii', 'ignore')
```
Comment (eefriedman) [inline python/servo/command_base.py]: It seems a little evil to assume that everyone's PATH is pure ASCII.

Comment (vvuk) [inline python/servo/command_base.py]: Well, I guess I could encode as utf-8 here instead.  Sometimes, and I don't know why, the PATH value gets munged and turned into a unicode string.  That can't be passed in as an environment variable.  utf-8 is probably a better option here.

Comment (eefriedman) [inline python/servo/command_base.py]: I don't really know enough about string encoding in Python to say what the right thing to do here is.

**Comment by vvuk** [comment]
Ok, I tracked down the unicode PATH junk.  virtualenv's activate_this.py script is doing that munging, because it's using some file functions that are somewhere returning a unicode string on windows (needlessly).  I added a comment and do the ascii-fying only on windows... I think doing ascii is as correct as anything else, because unicode env vars will simply not work with this anyway to begin with.

For the thread sleep bit, std::thread::sleep states that a signal won't cause an early return, whereas usleep() would get woken up by a signal... if std::thread::sleep is still okay, we can change that in a followup commit as it's not really windows specific.

**Comment by waddlesplash** [comment]
FWIW - the plan is to add a Windows buildbot so that checks are run on Windows for every commit, correct? Or is it too early for that?

**Comment by vvuk** [comment]
Yup, that's the plan.  I was going to work with someone to help make that happen, I just can't maintain it myself.

**Comment by waddlesplash** [comment]
Awesome. I don't use Linux anywhere but VMs anymore, and Rust on Haiku is still to alpha for Servo, so getting Servo working on Windows might motivate me to contribute...

**Comment by Jayflux** [comment]
@vvuk any way for me to help you with the windows buildbot?

**Comment by nox** [comment]
```
./components/util/workqueue.rs:13: use statement is not in alphabetical order
	expected: deque::{Abort, BufferPool, Data, Empty, Stealer, Worker}
	found: libc::funcs::posix88::unistd::usleep
./components/util/workqueue.rs:14: use statement is not in alphabetical order
	expected: libc::funcs::posix88::unistd::usleep
	found: deque::{Abort, BufferPool, Data, Empty, Stealer, Worker}
```

**Comment by pcwalton** [comment]
This looks good to me. Not sure who else should review—@jdm maybe?

**Comment by jdm** [comment]
Very little of this stuff appears to be in my area of expertise. @frewsxcv do you want to take a look at the python-related changes?

**Comment by frewsxcv** [comment]
FYI, there's a build error. Check out Travis

https://travis-ci.org/servo/servo/jobs/85117161

**Comment by vvuk** [comment]
Yes, this build can't succeed until the dependent pull requests are all merged in.

File: components/style/build.rs [hunk 3]
```diff
@@ -8,12 +8,32 @@ use std::io::Write;
...
+fn find_python() -> String {
+    if Command::new("python2.7").arg("--version").output().unwrap().status.success() {
         "python2.7"
     } else {
         "python"
+    }.to_owned()
+}
+
+fn main() {
+    let python = match env::var("PYTHON") {
```
Comment (frewsxcv) [inline components/style/build.rs]: Do we really need this here? It seems like if the user doesn't have Python 2.7 installed, they'll find out by `mach` anyways [here](https://github.com/servo/servo/blob/9fca41c7a7498cb6e27b914c45290dc4639d6f8b/python/mach_bootstrap.py#L90-L92).

File: python/servo/command_base.py [hunk 4]
```diff
@@ -17,6 +17,10 @@
 
 from mach.registrar import Registrar
 
+BIN_SUFFIX = ""
+if sys.platform == "win32":
+    BIN_SUFFIX = ".exe"
```
Comment (frewsxcv) [inline python/servo/command_base.py]: Alternatively:

```python
BIN_SUFFIX = ".exe" if sys.platform == "win32" else ""
```

**Comment by metajack** [comment]
Just to link them all together, this is blocked on:
- https://github.com/servo/skia/pull/74 
- https://github.com/servo/mozjs/pull/54 
- https://github.com/servo/glutin/pull/52

**Comment by frewsxcv** [comment]
Updated list:

* servo/skia#74
* servo/mozjs#54

**Comment by lygstate** [comment]
Updated list:

servo/skia#74

**Comment by Jayflux** [comment]
@lygstate looks like servo/skia#74 is done?

**Comment by lygstate** [comment]
wow looks great hope this can be merged now

**Comment by adamncasey** [comment]
I've had a go at rebasing (https://github.com/adamncasey/servo/tree/win32-rebase) and getting it to compile again.

A number of packages need cargo updating (but not too far, or else we hit use rustc-plugin; problems). Might be useful if these cargo updates are done in a separate PR.

However, I'm stuck on getting gaol and servo-glutin to build. https://gist.github.com/adamncasey/58da60a5672b62c9232e

servo-glutin could just be a cargo update or small change to fix.
But I can't see any indication of windows support in the gaol library - which could be the issue. @pcwalton ?

**Comment by adamncasey** [comment]
Fixed servo-glutin build: https://github.com/servo/glutin/pull/56

**Comment by adamncasey** [comment]
I stubbed out what's required for gaol to build on windows: https://github.com/adamncasey/gaol
I assume servo won't actually run on windows with this code, but getting it building on a CI would be a great start. Someone who understands gaol a bit more will likely want to rewrite everything I've done at some point.

I'm also finding that this commit to libANGLE broke the build on windows. Removing that one line fixes it.: https://github.com/ecoal95/angle/commit/429f752bab753443e0a942eba9a0bafc1ae3d66a
Quite strange. It causes dllexport / dllimport errors. 

The final problem I've hit is that I get linker errors at the end of the build:
```
note: C:\msys64\home\Adam\servo\servo\target\debug\deps\libjs-b63ddf1f18d33225.rlib(js-b63ddf1f18d33225.0.o): In function `js::jsapi_windows_gcc_64::JSTracer::setTracingLocation':
C:/msys64/home/Adam/servo/servo/.cargo\git\checkouts\rust-mozjs-ebb4917e843c0a11\master\src/jsapi_windows_gcc_64.rs:443: undefined reference to `JSTracer::setTracingLocation(void*)'
C:\msys64\home\Adam\servo\servo\target\debug\deps\libjs-b63ddf1f18d33225.rlib(js-b63ddf1f18d33225.0.o): In function `js::jsapi_windows_gcc_64::JSTracer::unsetTracingLocation':
C:/msys64/home/Adam/servo/servo/.cargo\git\checkouts\rust-mozjs-ebb4917e843c0a11\master\src/jsapi_windows_gcc_64.rs:447: undefined reference to `JSTracer::unsetTracingLocation()'
C:\msys64\home\Adam\servo\servo\target\debug\deps\libjs-b63ddf1f18d33225.rlib(js-b63ddf1f18d33225.0.o): In function `js::jsapi_windows_gcc_64::JSTracer::tracingLocation':
C:/msys64/home/Adam/servo/servo/.cargo\git\checkouts\rust-mozjs-ebb4917e843c0a11\master\src/jsapi_windows_gcc_64.rs:452: undefined reference to `JSTracer::tracingLocation(void**)'
collect2.exe: error: ld returned 1 exit status
```

This didn't happen when I built @vvuk 's code a few weeks back, so I'm a bit confused.

**Comment by vvuk** [comment]
This hacks around/fixes the angle issue:

https://github.com/vvuk/angle/commit/46a73e1b55287eb931420e181ab25c948af2f133

**Comment by vvuk** [comment]
I filed https://github.com/servo/mozjs/issues/70 for the JSTracer::setTracingLocation issue... gcc doesn't seem to be generating a public symbol for an inline-defined method.

**Comment by vvuk** [comment]
Using the msys python (see issue #9168) With the angle hack and the mozjs fix, and using the gaol linked above with win32 stubs, I get a complete build.

**Comment by larsbergstrom** [comment]
The angle issue has been merged in:
https://github.com/ecoal95/angle/pull/5

So, we now need to do the following to land:
- Land servo/mozjs#71
- Instead of changing gaol (https://github.com/adamncasey/gaol/commit/53a1af134df736ea4c980560b3bd52c3471f74e3), make the `-S` flag error out and `#[cfg(not(windows))]` around everything using gaol in Servo
- Add a Win32-Readme.md with a version of the instructions from https://hackpad.com/Servo-on-Windows-C1LPcI2bP25 and https://github.com/servo/servo/issues/9168

**Comment by larsbergstrom** [comment]
I've added the changes to cfg-out gaol and done a bunch of cargo updates here:
https://github.com/larsbergstrom/servo/tree/win32

We now just need to wait for mozjs to land and then to add a readme!

**Comment by Jayflux** [comment]
Im working on the readme, i need to start again though i think, some things have changed since i last ran through it all.

@vvuk do we still need to install the native python? (at the top of the hackpad)

Last night i hit the angle issue, is this now fixed on @larsbergstrom's win32 branch? or do i need to build of another branch?

**Comment by vvuk** [comment]
> @vvuk do we still need to install the native python? (at the top of the hackpad)

I haven't tried without it -- the mozjs makefile.cargo still explicitly references it.  I'll try without it shortly and see what happens.

**Comment by larsbergstrom** [comment]
@Jayflux my win32 branch does also have the angle update (it was landed, so it's just a cargo update). 

I'm in the middle of a small rebase because my hacky gaol fixes weren't split into a commit and didn't quite work well on some other platforms :-) I also need to make sure Cargo.lock is in sync across all the ports, since I did a bit of hand-editing on the main one just to get things to work. Should be up in ~30 minutes.

**Comment by larsbergstrom** [comment]
Just did another `push -f` to my win32 branch; this should have everything except for the mozjs fix and a readme.

**Comment by vvuk** [comment]
The native python seems to still be required; the issue seems to be that sys.executable is set incorrectly if python is invoked with an explicit c:/... path.

    vladimir@grizzly[467]$ pwd
    /c/proj/r/servo/target/debug/build/mozjs_sys-e1a009922d76a686
    vladimir@grizzly[468]$ C:/proj/r/servo/python/_virtualenv/bin/python2.7 -c 'import sys; print sys.executable'
    /c/proj/r/servo/target/debug/build/mozjs_sys-e1a009922d76a686/C:/proj/r/servo/python/_virtualenv/bin/python2.7

if it's invoked as /c/proj/..., then the path ends up correct.

**Comment by larsbergstrom** [comment]
moved to #9385

**Pull request closed without merging**